packages/ui/examples/next-openai/src/app/stream-ui/actions.tsx

```diff
@@ -0,0 +1,136 @@
+import { CoreMessage, generateId } from 'ai';
+import {
+  createAI,
+  createStreamableValue,
+  getMutableAIState as $getMutableAIState,
+  streamUI,
+} from 'ai/rsc';
+import { Message, BotMessage } from './message';
+import { z } from 'zod';
+import { run } from '@llamaindex/ui/rsc'
+import { OpenAI, OpenAIAgent } from 'llamaindex';
+
+type AIProviderNoActions = ReturnType<typeof createAI<AIState, UIState>>;
+// typed wrapper *without* actions defined to avoid circular dependencies
+const getMutableAIState = $getMutableAIState<AIProviderNoActions>;
+
+// mock function to fetch weather data
+const fetchWeatherData = async (location: string) => {
+  await new Promise(resolve => setTimeout(resolve, 1000));
+  return { temperature: '72°F' };
+};
+
+export async function submitUserMessage(content: string) {
+  'use server';
+
+  const aiState = getMutableAIState();
+
+  aiState.update({
+    ...aiState.get(),
+    messages: [
+      ...aiState.get().messages,
+      { id: generateId(), role: 'user', content },
+    ],
+  });
+
+  let textStream: undefined | ReturnType<typeof createStreamableValue<string>>;
+  let textNode: React.ReactNode;
+  const agent = new OpenAIAgent({
+    llm: new OpenAI({
+      model: 'gpt-4-turbo'
+    }),
+    systemPrompt: 'You are a weather assistant.',
+    tools: [
+			{
+				name: 'get_current_weather',
+				description: 'Get the current weather',
+				parameters: z.object({
+					location: z.string(),
+				}),
+			},
+    ]
+  })
+  const result =  run(() => {
```

Comment:
@himself65, how would that from the caller side look like? I didn't look into the problem, but I think from the DX we're closest to OpenAI's assistants: https://sdk.vercel.ai/docs/ai-sdk-ui/openai-assistants maybe we can provide our own `useAssistants` function?